File: Example/CoreData/SyncKitCoreDataExampleTests/QSObject.m
//
//  QSObject.m
//  QSCloudKitSynchronizer
//
//  Created by Manuel Entrena on 24/07/2016.
//  Copyright © 2016 Manuel. All rights reserved.
//

#import "QSObject.h"
#import <CloudKit/CloudKit.h>

@implementation QSObject

- (instancetype)initWithIdentifier:(NSString *)identifier number:(NSNumber *)number
{
    self = [super init];
    if (self) {
        self.identifier = identifier;
        self.number = number;
    }
    return self;
}

- (instancetype)initWithRecord:(CKRecord *)record
{
    return [self initWithIdentifier:record.recordID.recordName number:record[@"number"]];
}

- (CKRecord *)recordWithZoneID:(CKRecordZoneID *)zoneID
{
    CKRecordID *recordID = [[CKRecordID alloc] initWithRecordName:self.identifier zoneID:zoneID];
    CKRecord *record = [[CKRecord alloc] initWithRecordType:@"object" recordID:recordID];
    record[@"number"] = self.number;
    return record;
}

- (CKRecordID *)recordIDWithZoneID:(CKRecordZoneID *)zoneID
{
    return [[CKRecordID alloc] initWithRecordName:self.identifier zoneID:zoneID];
}

- (void)saveRecord:(CKRecord *)record
{
    self.number = record[@"number"];
}

@end

File: Example/CoreData/SyncKitCoreDataExample/SyncKitCoreDataExampleTests/Mocks/QSObject.swift
//
//  QSObject.swift
//  SyncKitCoreDataExampleTests
//
//  Created by Manuel Entrena on 12/06/2019.
//  Copyright © 2019 Manuel Entrena. All rights reserved.
//

import Foundation
import SyncKit
import CloudKit

class QSObject: Codable, Equatable {
    static func == (lhs: QSObject, rhs: QSObject) -> Bool {
        return lhs.identifier == rhs.identifier
    }
    
    var number: Int?
    var identifier: String
    
    init(identifier: String, number: Int?) {
        self.identifier = identifier
        self.number = number
    }
    
    convenience init(record: CKRecord) {
        self.init(identifier: record.recordID.recordName, number: record["number"])
    }
    
    func record(with zoneID: CKRecordZone.ID) -> CKRecord {
        let record = CKRecord(recordType: "object", recordID: CKRecord.ID(recordName: identifier, zoneID: zoneID))
        record["number"] = number
        return record
    }
    
    func recordID(zoneID: CKRecordZone.ID) -> CKRecord.ID {
        return CKRecord.ID(recordName: identifier, zoneID: zoneID)
    }
    
    func save(record: CKRecord) {
        number = record["number"]
    }
}
